nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -446,6 +469,17 @@ where
                         .await;
                     }
 
+                    Some(mixed_msg) = incoming_mixed_msgs.next() => {
+                        match wire::deserialize::<Block<ClPool::Item, DaPool::Item>>(&mixed_msg) {
+                            Ok(block) => {
+                                adapter.broadcast(NetworkMessage::Block(block)).await;
+                            },
+                            Err(e) => {
+                                tracing::error!("Error deserializing mixed message: {e}");
+                            },
+                        }
+                    }
+
```

Comment:
Why is this in consensus?
I think this is wrong. We are not validating anything in the block yet. We are just piping to somewhere else. This definitely should be done by the mixnet service itself. Consensus should just read from the public network and write to the mixntet (?)

Reply:
I was also thinking about this when implementing this code. To be clear, like the network adapter that returns a stream of `Block`, do you mean that it would be better to make the mix adapter return a stream of `Block` (not a stream of `Vec<u8>`)?

Reply:
> I was also thinking about this when implementing this code. To be clear, like the network adapter that returns a stream of `Block`, do you mean that it would be better to make the mix adapter return a stream of `Block` (not a stream of `Vec<u8>`)?

Uhm, from the mixnet perspective its fine its just bytes. But what I mean is that this relaying and deserializing should not be done in consensus at all.

Reply:
I think I don't catch your point correctly. Do you think that reading messages from the mix network and sending them to the broadcast network should be done inside the mix service? The mix service implements only the mix network. And, the user of the mix service (i.e. consensus service) can choose what to do with messages returned from the mix network.

Reply:
Ah, okay. We don't need to stick to the following sentence that I wrote:
> The mix service implements only the mix network.

We can make the mix service use the broadcast service. But, I'd like to check if you're talking about this point.

Reply:
Okay, I did a sync with @madxor . I'll restructure the services like below:
- Consensus service listens messages from the network service (the broadcast network, in a term familiar to Marcin).
- Consensus service sends messages to the mix service.
- Mix service wraps the messages and propagate them through the mix network.
**- Mix service sends fully-unwrapped messages to the network service.**

The bold one is what @danielSanchezQ is pointing out in this thread, if I understand him correctly. That's right. I'll update the implementation accordingly.

Reply:
Yeah, this is exactly the flow I meant. I would go for it!

Reply:
If I get it correctly, in case of other services using mixnet for other types of messages would use different kind of mixnet adapter? This would allow to have one mixnet service and multiple adapters that would take care of filtering and deserialization.

In this case MixNetworkAdapter is already defined in the consensus module, so I assume it will be only used for consensus, thus could handle the deserialization as well.

Reply:
> If I get it correctly, in case of other services using mixnet for other types of messages would use different kind of mixnet adapter? This would allow to have one mixnet service and multiple adapters that would take care of filtering and deserialization.
> 
> In this case MixNetworkAdapter is already defined in the consensus module, so I assume it will be only used for consensus, thus could handle the deserialization as well.

yes it would be better to handle that in the adapter itself. But in this case consensus just writes/sends to mixnet, but it does not read to it. So instead of deserialization the adapter would probably just serialize.

Reply:
Yeah.. At the beginning, I imagines the case which uses the mix service but not the network service. That's why I clearly decoupled the mix service from the network service. But, if that is not the case that we should consider, I can couple the mix service and the network service..

Reply:
> Mix service sends fully-unwrapped messages to the network service.

I still think this should be handled in the consensus mixnet adapter, as it is consensus related - the block needs to be broadcast in a specific topic, that libp2p network adapter knows about. Mixnet adapter in consensus could require a libp2p network adapter when initializing it.

> But, if that is not the case that we should consider, I can couple the mix service and the network service

We could still keep services uncoupled and agnostic to the details that consensus requires when two adapters in consensus?

Reply:
> the block needs to be broadcast in a specific topic, that libp2p network adapter knows about. Mixnet adapter in consensus could require a libp2p network adapter when initializing it.

Ah, true. That's what the mix service doesn't know. I think you're right. Let me try to implement it and check which way is the most reasonable.

Reply:
> > Mix service sends fully-unwrapped messages to the network service.
> 
> I still think this should be handled in the consensus mixnet adapter, as it is consensus related - the block needs to be broadcast in a specific topic, that libp2p network adapter knows about. Mixnet adapter in consensus could require a libp2p network adapter when initializing it.
> 
> > But, if that is not the case that we should consider, I can couple the mix service and the network service
> 
> We could still keep services uncoupled and agnostic to the details that consensus requires when two adapters in consensus?

I understand this. But take into consideration that we would be using consensus as a bridge for networking **when is not needed**. For consensus just 2 things matter:
* Receive blocks 
* Sends blocks

It should not bridge two network services just because it understands them. So either we merge the two adapter into a single one or this is dealt in mixnet-service level. Probably the key question here is *is the broadcasting part of the mixnet protocol?* if yes, it can be handle there, if not in the adapter. But consensus should just have to calls to adapter to send and receive **full decoded blocks**.

Reply:
> Probably the key question here is is the broadcasting part of the mixnet protocol?

It just has been agreed to as YES, in the spec.

Reply:
Resolved by #829